posthog/hogql_queries/lifecycle_query_runner.py

```diff
@@ -0,0 +1,252 @@
+            "events_query": self.events_query,
+            "periods_query": self.periods_query,
+        }
+        with self.timings.measure("lifecycle_query"):
+            lifecycle_query = parse_select(
+                """
+                    SELECT groupArray(start_of_period) AS date,
+                           groupArray(counts) AS total,
+                           status
+                    FROM (
+                        SELECT
+                            status = 'dormant' ? negate(sum(counts)) : negate(negate(sum(counts))) as counts,
+                            start_of_period,
+                            status
+                        FROM (
+                            SELECT
+                                periods.start_of_period as start_of_period,
+                                0 AS counts,
+                                status
+                            FROM {periods_query} as periods
+                            CROSS JOIN (
+                                SELECT status
+                                FROM (SELECT 1)
+                                ARRAY JOIN ['new', 'returning', 'resurrecting', 'dormant'] as status
+                            ) as sec
+                            ORDER BY status, start_of_period
+                            UNION ALL
+                            SELECT
+                                start_of_period, count(DISTINCT person_id) AS counts, status
+                            FROM {events_query}
+                            GROUP BY start_of_period, status
+                        )
+                        WHERE start_of_period <= dateTrunc({interval}, {date_to})
+                            AND start_of_period >= dateTrunc({interval}, {date_from})
+                        GROUP BY start_of_period, status
+                        ORDER BY start_of_period ASC
+                    )
+                    GROUP BY status
+                """,
+                placeholders,
+                timings=self.timings,
+            )
+        return lifecycle_query
+
+    def to_persons_query(self) -> str:
+        # TODO: add support for selecting and filtering by breakdowns
+        with self.timings.measure("persons_query"):
+            return parse_select(
+                """
+                SELECT
+                    person_id, start_of_period as breakdown_1, status as breakdown_2
+                FROM
+                    {events_query}
+                """,
+                placeholders={"events_query": self.events_query},
+            )
+
+    def run(self) -> LifecycleQueryResponse:
+        response = execute_hogql_query(
+            query_type="LifecycleQuery",
+            query=self.to_query(),
+            team=self.team,
+            timings=self.timings,
+        )
+
+        # TODO: can we move the data conversion part into the query as well? It would make it easier to swap
+        # e.g. the LifecycleQuery with HogQLQuery, while keeping the chart logic the same.
```

Comment:
Handling the response mangling on the app layer seems the right place to me 🤔

Reply:
It does... but if we want to be able to move towards a flexible system where we can swap out `LifecycleQuery` with `HogQLQuery` at will, we'll need a layer that adds these custom transformations on top.

The options we have are
a) Avoid the problem by having everything in SQL (might not be possible/practical... and how will `InsightVizNode` know what to render if we just give a query?) 
b) `{ kind: 'LifecycleQuery', source: { ...HogQLQuery } }`
c) `{ kind: 'InsightDataSourceQuery', type: 'Lifecycle', source: { ...HogQLQuery }  }` as we need to tell the `InsightVizNode` _somehow_ what to render
d) ...? 🤔